File: src/main/java/com/freya02/botcommands/api/components/ComponentListener.java
package com.freya02.botcommands.api.components;

import com.freya02.botcommands.api.ExceptionHandler;
import com.freya02.botcommands.api.Logging;
import com.freya02.botcommands.api.components.event.ButtonEvent;
import com.freya02.botcommands.api.components.event.SelectionEvent;
import com.freya02.botcommands.api.parameters.ComponentParameterResolver;
import com.freya02.botcommands.internal.BContextImpl;
import com.freya02.botcommands.internal.RunnableEx;
import com.freya02.botcommands.internal.application.CommandParameter;
import com.freya02.botcommands.internal.components.ComponentDescriptor;
import com.freya02.botcommands.internal.utils.Utils;
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent;
import net.dv8tion.jda.api.events.interaction.component.GenericComponentInteractionCreateEvent;
import net.dv8tion.jda.api.events.interaction.component.SelectMenuInteractionEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import net.dv8tion.jda.api.hooks.SubscribeEvent;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.function.Consumer;
import java.util.function.Function;

public class ComponentListener extends ListenerAdapter {
	private static final Logger LOGGER = Logging.getLogger();

	private final ExecutorService idHandlingExecutor = Executors.newSingleThreadExecutor(r -> {
		final Thread thread = new Thread(r);
		thread.setDaemon(false);
		thread.setUncaughtExceptionHandler((t, e) -> Utils.printExceptionString("An unexpected exception happened in a component ID handler thread '" + t.getName() + "':", e));
		thread.setName("Component ID handling thread");

		return thread;
	});

	private final BContextImpl context;
	private final ComponentManager componentManager;

	private final Map<String, ComponentDescriptor> buttonsMap;
	private final Map<String, ComponentDescriptor> selectionMenuMap;

	private int callbackThreadNumber;
	private final ExecutorService callbackExecutor = Utils.createCommandPool(r -> {
		final Thread thread = new Thread(r);
		thread.setDaemon(false);
		thread.setUncaughtExceptionHandler((t, e) -> Utils.printExceptionString("An unexpected exception happened in a component callback thread '" + t.getName() + "':", e));
		thread.setName("Component callback thread #" + callbackThreadNumber++);

		return thread;
	});

	public ComponentListener(BContextImpl context, Map<String, ComponentDescriptor> buttonsMap, Map<String, ComponentDescriptor> selectionMenuMap) {
		this.context = context;
		this.componentManager = Utils.getComponentManager(context);
		this.buttonsMap = buttonsMap;
		this.selectionMenuMap = selectionMenuMap;

		Components.setContext(context);
	}

	@SubscribeEvent
	@Override
	public void onGenericComponentInteractionCreate(@NotNull GenericComponentInteractionCreateEvent event) {
		if (!(event instanceof ButtonInteractionEvent) && !(event instanceof SelectMenuInteractionEvent)) return;

		runHandler(() -> handleComponentInteraction(event), event);
	}

	private void handleComponentInteraction(@NotNull GenericComponentInteractionCreateEvent event) throws Exception {
		for (ComponentInteractionFilter componentFilter : context.getComponentFilters()) {
			if (!componentFilter.isAccepted(new ComponentFilteringData(context, event))) {
				LOGGER.trace("Cancelled component interaction due to filter");

				return;
			}
		}

		try (FetchResult fetchResult = componentManager.fetchComponent(event.getComponentId())) {
			final FetchedComponent fetchedComponent = fetchResult.getFetchedComponent();

			if (fetchedComponent == null) {
				event.reply(context.getDefaultMessages(event).getComponentNotFoundErrorMsg())
						.setEphemeral(true)
						.queue();

				return;
			}

			final ComponentType idType = fetchedComponent.getType();
			if ((idType == ComponentType.PERSISTENT_BUTTON || idType == ComponentType.LAMBDA_BUTTON) && !(event instanceof ButtonInteractionEvent)) {
				LOGGER.error("Received a button id type but event is not a ButtonInteractionEvent");

				return;
			}

			if ((idType == ComponentType.PERSISTENT_SELECTION_MENU || idType == ComponentType.LAMBDA_SELECTION_MENU) && !(event instanceof SelectMenuInteractionEvent)) {
				LOGGER.error("Received a selection menu id type but event is not a SelectMenuInteractionEvent");

				return;
			}

			switch (idType) {
				case PERSISTENT_BUTTON -> componentManager.handlePersistentButton(event,
						fetchResult,
						e -> onError(event, e),
						data -> runCallback(() -> handlePersistentComponent(event,
										buttonsMap,
										data.getHandlerName(),
										data.getArgs(),
										descriptor -> new ButtonEvent(descriptor.getMethod(), context, (ButtonInteractionEvent) event)),
								event));
				case LAMBDA_BUTTON -> componentManager.handleLambdaButton(event,
						fetchResult,
						e -> onError(event, e),
						data -> runCallback(() -> data.getConsumer().accept(new ButtonEvent(null, context, (ButtonInteractionEvent) event)), event)
				);
				case PERSISTENT_SELECTION_MENU -> componentManager.handlePersistentSelectMenu(event,
						fetchResult,
						e -> onError(event, e),
						data -> runCallback(() -> handlePersistentComponent(event,
										selectionMenuMap,
										data.getHandlerName(),
										data.getArgs(),
										descriptor -> new SelectionEvent(descriptor.getMethod(), context, (SelectMenuInteractionEvent) event)),
								event));
				case LAMBDA_SELECTION_MENU -> componentManager.handleLambdaSelectMenu(event,
						fetchResult,
						e -> onError(event, e),
						data -> runCallback(() -> data.getConsumer().accept(new SelectionEvent(null, context, (SelectMenuInteractionEvent) event)), event));
				default -> throw new IllegalArgumentException("Unknown id type: " + idType.name());
			}
		}
	}

	private void runHandler(RunnableEx code, @NotNull GenericComponentInteractionCreateEvent event) {
		idHandlingExecutor.execute(() -> {
			try {
				long start = System.nanoTime();
				code.run();
				long end = System.nanoTime();

				LOGGER.trace("Component handler took {} ms", (end - start) / 1000000.0);
			} catch (Throwable e) {
				final ExceptionHandler handler = context.getUncaughtExceptionHandler();
				if (handler != null) {
					handler.onException(context, event, e);

					return;
				}

				Throwable baseEx = Utils.getException(e);

				Utils.printExceptionString("Unhandled exception in thread '" + Thread.currentThread().getName() + "' while executing the component ID handler", baseEx);
				if (event.isAcknowledged()) {
					event.getHook().sendMessage(context.getDefaultMessages(event).getGeneralErrorMsg()).setEphemeral(true).queue();
				} else {
					event.reply(context.getDefaultMessages(event).getGeneralErrorMsg()).setEphemeral(true).queue();
				}

				context.dispatchException("Exception in component ID handler", baseEx);
			}
		});
	}

	private void runCallback(RunnableEx code, @NotNull GenericComponentInteractionCreateEvent event) {
		callbackExecutor.execute(() -> {
			try {
				long start = System.nanoTime();
				code.run();
				long end = System.nanoTime();

				LOGGER.trace("Component callback took {} ms", (end - start) / 1000000.0);
			} catch (Throwable e) {
				final ExceptionHandler handler = context.getUncaughtExceptionHandler();
				if (handler != null) {
					handler.onException(context, event, e);

					return;
				}

				Throwable baseEx = Utils.getException(e);

				Utils.printExceptionString("Unhandled exception in thread '" + Thread.currentThread().getName() + "' while executing a component callback", baseEx);
				if (event.isAcknowledged()) {
					event.getHook().sendMessage(context.getDefaultMessages(event).getGeneralErrorMsg()).setEphemeral(true).queue();
				} else {
					event.reply(context.getDefaultMessages(event).getGeneralErrorMsg()).setEphemeral(true).queue();
				}

				context.dispatchException("Exception in component callback", baseEx);
			}
		});
	}

	private void handlePersistentComponent(GenericComponentInteractionCreateEvent event,
	                                       Map<String, ComponentDescriptor> map,
	                                       String handlerName,
	                                       String[] args,
	                                       Function<ComponentDescriptor, ? extends GenericComponentInteractionCreateEvent> eventFunction) {
		final ComponentDescriptor descriptor = map.get(handlerName);

		if (descriptor == null) {
			LOGGER.error("No component descriptor found for component handler '{}'", handlerName);

			return;
		}

		final var parameters = descriptor.getParameters();
		if (parameters.getOptionCount() != args.length) {
			throw new IllegalArgumentException("Resolver for %s has %d arguments but component had %d data objects".formatted(Utils.formatMethodShort(descriptor.getMethod()), parameters.size(), args.length));
		}

		final Consumer<Throwable> throwableConsumer = getThrowableConsumer(handlerName, args);
		try {
			//For some reason using an array list instead of a regular array
			// magically unboxes primitives when passed to Method#invoke
			final List<Object> methodArgs = new ArrayList<>(parameters.size() + 1);

			methodArgs.add(eventFunction.apply(descriptor));

			int optionIndex = 0;
			for (final CommandParameter<ComponentParameterResolver> parameter : parameters) {
				final Object obj;
				if (parameter.isOption()) {
					final String arg = args[optionIndex];
					optionIndex++;

					obj = parameter.getResolver().resolve(context, descriptor, event, arg);

					if (obj == null) {
						throw new IllegalArgumentException("Component id '%s', tried to resolve '%s' with an option resolver %s on method %s but result is null".formatted(
								event.getComponentId(),
								arg,
								parameter.getCustomResolver().getClass().getSimpleName(),
								Utils.formatMethodShort(descriptor.getMethod())
						));
					}
				} else {
					obj = parameter.getCustomResolver().resolve(context, descriptor, event);

					if (obj == null) {
						throw new IllegalArgumentException("Component id '%s', tried to use custom resolver %s on method %s but result is null".formatted(
								event.getComponentId(),
								parameter.getCustomResolver().getClass().getSimpleName(),
								Utils.formatMethodShort(descriptor.getMethod())
						));
					}
				}

				methodArgs.add(obj);
			}

			descriptor.getMethodRunner().invoke(methodArgs.toArray(), throwableConsumer);
		} catch (Exception e) {
			throwableConsumer.accept(e);
		}
	}

	@NotNull
	private Consumer<Throwable> getThrowableConsumer(String handlerName, String[] args) {
		return e -> LOGGER.error("An exception occurred while handling a persistent component '{}' with args {}", handlerName, Arrays.toString(args), e);
	}

	private void onError(GenericComponentInteractionCreateEvent event, ComponentErrorReason reason) {
		event.reply(reason.getReason(context.getDefaultMessages(event)))
				.setEphemeral(true)
				.queue();
	}
}

File: src/main/kotlin/com/freya02/botcommands/internal/components/ComponentListener.kt
package com.freya02.botcommands.internal.components

import com.freya02.botcommands.api.Logging
import com.freya02.botcommands.api.components.ComponentErrorReason
import com.freya02.botcommands.api.components.ComponentFilteringData
import com.freya02.botcommands.api.components.ComponentManager
import com.freya02.botcommands.api.components.ComponentType
import com.freya02.botcommands.api.components.event.ButtonEvent
import com.freya02.botcommands.api.components.event.SelectionEvent
import com.freya02.botcommands.core.api.annotations.BEventListener
import com.freya02.botcommands.core.api.annotations.BService
import com.freya02.botcommands.internal.BContextImpl
import com.freya02.botcommands.internal.getDeepestCause
import com.freya02.botcommands.internal.parameters.CustomMethodParameter
import com.freya02.botcommands.internal.parameters.MethodParameterType
import com.freya02.botcommands.internal.throwInternal
import com.freya02.botcommands.internal.throwUser
import com.freya02.botcommands.internal.utils.ReflectionUtilsKt.shortSignature
import dev.minn.jda.ktx.messages.reply_
import dev.minn.jda.ktx.messages.send
import kotlinx.coroutines.launch
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent
import net.dv8tion.jda.api.events.interaction.component.GenericComponentInteractionCreateEvent
import net.dv8tion.jda.api.events.interaction.component.SelectMenuInteractionEvent
import kotlin.reflect.KParameter
import kotlin.reflect.full.callSuspendBy
import kotlin.reflect.full.instanceParameter
import kotlin.reflect.full.valueParameters

private val LOGGER = Logging.getLogger()

@BService
internal class ComponentListener(
    private val context: BContextImpl,
    private val componentsHandlerContainer: ComponentsHandlerContainer,
    private val componentsManager: ComponentManager
) {
    @BEventListener
    internal fun onButtonClick(event: ButtonInteractionEvent) {
        onComponentInteraction(event)
    }

    @BEventListener
    internal fun onSelectMenuInteraction(event: SelectMenuInteractionEvent) {
        onComponentInteraction(event)
    }

    private fun onComponentInteraction(event: GenericComponentInteractionCreateEvent) {
        try {
            for (componentFilter in context.componentFilters) {
                if (!componentFilter.isAccepted(ComponentFilteringData(context, event))) {
                    return
                }
            }

            val scope = context.config.coroutineScopesConfig.componentsScope

            componentsManager.fetchComponent(event.componentId) { fetchResult ->
                val fetchedComponent = fetchResult.fetchedComponent ?: run {
                    event.reply_(context.getDefaultMessages(event).componentNotFoundErrorMsg, ephemeral = true).queue()

                    return@fetchComponent
                }

                val idType = fetchedComponent.type
                if ((idType == ComponentType.LAMBDA_BUTTON || idType == ComponentType.PERSISTENT_BUTTON) && event !is ButtonInteractionEvent) {
                    throwInternal("Received a button id type but event is not a ButtonInteractionEvent")
                }

                if ((idType == ComponentType.PERSISTENT_SELECTION_MENU || idType == ComponentType.LAMBDA_SELECTION_MENU) && event !is SelectMenuInteractionEvent) {
                    throwInternal("Received a selection menu id type but event is not a SelectMenuInteractionEvent")
                }

                when (idType) {
                    ComponentType.PERSISTENT_BUTTON -> {
                        componentsManager.handlePersistentButton(event, fetchResult, { onError(event, it) }) {
                            val descriptor = componentsHandlerContainer.getButtonDescriptor(it.handlerName)
                                ?: throwUser("No component descriptor found for component handler '${it.handlerName}'")

                            scope.launch {
                                handlePersistentComponent(event, descriptor, it.args) {
                                    ButtonEvent(descriptor.method, context, event as ButtonInteractionEvent)
                                }
                            }
                        }
                    }
                    ComponentType.LAMBDA_BUTTON -> {
                        componentsManager.handleLambdaButton(event, fetchResult, { onError(event, it) }) {
                            scope.launch { //The scope is used for consuming the event asynchronously, as to free the ongoing transaction
                                it.consumer.accept(ButtonEvent(null, context, event as ButtonInteractionEvent))
                            }
                        }
                    }
                    ComponentType.PERSISTENT_SELECTION_MENU -> {
                        componentsManager.handlePersistentSelectMenu(event, fetchResult, { onError(event, it) }) {
                            val descriptor = componentsHandlerContainer.getSelectMenuDescriptor(it.handlerName)
                                ?: throwUser("No component descriptor found for component handler '${it.handlerName}'")

                            scope.launch {
                                handlePersistentComponent(event, descriptor, it.args) {
                                    SelectionEvent(descriptor.method, context, event as SelectMenuInteractionEvent)
                                }
                            }
                        }
                    }
                    ComponentType.LAMBDA_SELECTION_MENU -> {
                        componentsManager.handleLambdaSelectMenu(event, fetchResult, { onError(event, it) }) {
                            scope.launch { //The scope is used for consuming the event asynchronously, as to free the ongoing transaction
                                it.consumer.accept(SelectionEvent(null, context, event as SelectMenuInteractionEvent))
                            }
                        }
                    }
                }
            }
        } catch (e: Throwable) {
            handleException(e, event)
        }
    }

    private fun handleException(e: Throwable, event: GenericComponentInteractionCreateEvent) {
        val handler = context.uncaughtExceptionHandler
        if (handler != null) {
            handler.onException(context, event, e)
            return
        }

        val baseEx = e.getDeepestCause()

        LOGGER.error(
            "Unhandled exception in thread '${Thread.currentThread().name}' while executing a component", baseEx
        )

        val generalErrorMsg = context.getDefaultMessages(event).generalErrorMsg
        when {
            event.isAcknowledged -> event.hook.send(generalErrorMsg).queue()
            else -> event.reply_(generalErrorMsg, ephemeral = true).queue()
        }

        context.dispatchException("Exception in component handler", baseEx)
    }

    private suspend fun handlePersistentComponent(
        event: GenericComponentInteractionCreateEvent,
        descriptor: ComponentDescriptor,
        buttonArgs: Array<out String>,
        eventSupplier: () -> GenericComponentInteractionCreateEvent
    ) {
        val parameters = descriptor.parameters
        if (parameters.optionCount != buttonArgs.size) {
            throwUser("Resolver for ${descriptor.method.shortSignature} has ${parameters.optionCount} arguments but component had ${buttonArgs.size} arguments")
        }

        val objects: MutableMap<KParameter, Any?> = hashMapOf()
        objects[descriptor.method.instanceParameter!!] = descriptor.instance
        objects[descriptor.method.valueParameters.first()] = eventSupplier()

        var optionIndex = 0
        parameters.forEach { parameter ->
            when (parameter.methodParameterType) {
                MethodParameterType.COMMAND -> {
                    parameter as ComponentHandlerParameter

                    val buttonArg = buttonArgs[optionIndex++]
                    val resolved = parameter.resolver.resolve(context, descriptor, event, buttonArg)
                        ?: throwUser("Component id ${event.componentId}, tried to resolve '${buttonArg}' with option resolver ${parameter.resolver.javaClass.simpleName} on method ${descriptor.method.shortSignature} but result was null")

                    objects[parameter.kParameter] = resolved
                }
                MethodParameterType.CUSTOM -> {
                    parameter as CustomMethodParameter

                    val resolved = parameter.resolver.resolve(context, descriptor, event)
                        ?: throwUser("Component id ${event.componentId}, tried to resolve custom option with ${parameter.resolver.javaClass.simpleName} on method ${descriptor.method.shortSignature} but result was null")

                    objects[parameter.kParameter] = resolved
                }
                else -> TODO()
            }
        }

        descriptor.method.callSuspendBy(objects)
    }

    private fun onError(event: GenericComponentInteractionCreateEvent, reason: ComponentErrorReason) {
        event.reply_(reason.getReason(context.getDefaultMessages(event)), ephemeral = true).queue()
    }
}